### PR TITLE
Robust Toolbox Launcher Redesign

### PR DESCRIPTION
**Pull request by ike709**
Read the markdown file.

Supersedes #117 in my opinion (I didn't actually write any of that nor directly incorporate it into this, that's all ZeWaka).

### CODE REVIEW

**Comment by ike709** [comment]
To clarify something on Discord (I can refine it in the doc later), the "Robust Launcher" won't support multiple hubs. The reason the diagram has hubs other than RobustHub is:

1. Inevitable Spite Hub is an example of a _hard fork_ of the launcher and all other infrastructure. This is comparable to the existing group of people who want to fork all of the infra and do their own thing.
2. Johnny GameDeveloper is an example of using Robust Toolbox like any other game engine (e.g. Unity) to develop and release a game. This game would want its own infrastructure rather than being part of RobustHub.

File: src/en/proposals/ike709-robusthub.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+#### Adding Alternate Server Hubs
+On the page for a specific game, there will be a button to configure the Server Hubs for that game (assuming the game allows alternate hubs in the RobustHub metadata).
+
+#### What About Other Game Hubs?
+Whether or not other Game Hubs should be addable to Robust Launcher is at the discretion of RT project managers. If they opt not to allow this functionality, other Game Hubs will ship their own launcher forks pointed at their hub instead of RobustHub.
+
+#### Optional: Steam Release
+I'm of the opinion that Robust Toolbox should have its own dedicated Steam page which ships Robust Launcher directly, to reduce confusion when people want to play other RT projects like OpenDream via Steam. I'm not saying we should do this *immediately*, but I think it's worth serious consideration if any not-SS14-adjacent projects are created in Robust Toolbox.
+
+We should also check Steam's policy for this sort of thing, as the SS14 launcher at this point would just be Robust Launcher in Game-specific Mode for SS14 (see below).
```
Comment (deathride58) [inline src/en/proposals/ike709-robusthub.md:70]: IRT Steam's policy: Trackmania 2 (consisting of Canyon, Valley, Stadium, and Lagoon) and Shootmania collectively serve as existing precedent. None of these individual titles on Steam contain a full respective game, just the Maniaplanet launcher (which handles downloading and installing the actual game), with the title on Steam just quickstarting into a specific title via launch arg. Backing out of an individual title's main menu does not exit the game, but rather, puts you in Maniaplanet's home screen, regardless of whether or not it's the only Maniaplanet title you own.

We feel like game-specific mode isn't entirely necessary at all due to this precedent. Probably the only thing that'll be necessary is that the game merely boots into the given hub by default.

**Comment by ike709** [comment]
tl;dr summary that I posted on GitHub:

image

File: src/en/proposals/ike709-robusthub.md (hunk 1 above)
Comment (ike709) [reply src/en/proposals/ike709-robusthub.md:70]: The reason "Game-specific Mode" exists at all is because in prior discussions PJB outright refused to have a BYOND-style launcher that opened into a game list because it would lead to player confusion when they launch the Steam game "Space Station 14" and get a list of other games.

**Comment by ike709** [comment]
Added a second diagram showing how launchers fit into things.

**Comment by ike709** [comment]
Added additional clarification to the Game Hubs section.

File: src/en/proposals/ike709-robusthub.md (hunk 1 above)
Comment (PJB3005) [reply src/en/proposals/ike709-robusthub.md:67]: I mean literally gmod.

**Comment by PJB3005** [comment]
Don't like the idea that server hubs have to be per-game. Sounds easier if there was just an easy field to filter the game name in the advertisement data.

File: src/en/proposals/ike709-robusthub.md (hunk 1 above)
Comment (ike709) [reply src/en/proposals/ike709-robusthub.md:67]: @PJB3005 Most gmod games don't have their own Steam store pages.

**Comment by ike709** [comment]
Adjusted Server Hubs to support multiple games per hub.

**Comment by ike709** [comment]
Added an optional section to run a game jam once this system is fully implemented, see latest commit for details.

**Comment by Jezithyr** [comment]
This is a pretty great proposal, as a design I think it's a great next step in the future of robust. Byond2 electric boogaloo!

**Comment by Jezithyr** [comment]
Eventually it would also be neat to allow developers to "bring their own game engine" and use robust for handling accounts, sessions and server browsing. But that would very much be a wyci feature.
